File: go/view_test.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

package aontu


import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestViewTreeUnparseableSource(t *testing.T) {
	// An unclosed STRING does not parse, which is what reaches
	// parseEntry's error return.
	r := New().ViewTree("a: \"unterminated", nil)
	if "error" != r.Verdict || "tree" != r.Kind {
		t.Fatalf("verdict %q kind %q, want error tree", r.Verdict, r.Kind)
	}
	if 1 != len(r.Errors) || "" == r.Errors[0].Code {
		t.Fatalf("errors: %v", r.Errors)
	}
	if nil != r.Text {
		t.Fatalf("a refusal carried a figure: %q", *r.Text)
	}
}

// A NIL ROOT WITH AN EMPTY ERROR LIST (use-cases/BUGS.md 43): the
// id-spread refusal IS the root, so ctx.err is empty and the finding
// is built from the root itself.
func TestViewTreeNilRootWithNoCollectedError(t *testing.T) {
	r := New().ViewTree("&:\n", &ViewOptions{})
	if "error" != r.Verdict {
		t.Fatalf("verdict %q, want error", r.Verdict)
	}
	if 1 != len(r.Errors) || "" == r.Errors[0].Code {
		t.Fatalf("errors: %v", r.Errors)
	}
}

// A rendered figure is present even when empty: a model with no links
// draws nothing, honestly, rather than refusing.
func TestViewTreeOfNoEdgesIsAnEmptyFigure(t *testing.T) {
	r := New().ViewTree("a: 1\n", nil)
	if "rendered" != r.Verdict || nil == r.Text || "" != *r.Text {
		t.Fatalf("report: %+v", r)
	}
}


func viewTempDoc(t *testing.T, dir, name, src string) string {
	t.Helper()
	file := filepath.Join(dir, name)
	if err := os.MkdirAll(filepath.Dir(file), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(file, []byte(src), 0o600); err != nil {
		t.Fatal(err)
	}
	return file
}

func TestViewDocName(t *testing.T) {
	for _, c := range []struct{ file, entry, want string }{
		{"", "", "-"},
		{"", "dir/entry.aon", "entry.aon"},
		{"dir/entry.aon", "dir/entry.aon", "entry.aon"},
		{"lib/x.aon", "dir/entry.aon", "lib/x.aon"},
		{"/abs/x.aon", "", "/abs/x.aon"},
	} {
		if got := viewDocName(c.file, c.entry); got != c.want {
			t.Fatalf("viewDocName(%q, %q) = %q, want %q", c.file, c.entry, got, c.want)
		}
	}
	abs := filepath.Join(t.TempDir(), "sub", "inc.aon")
	entry := filepath.Join(filepath.Dir(filepath.Dir(abs)), "entry.aon")
	if got := viewDocName(abs, entry); filepath.Join("sub", "inc.aon") != got {
		t.Fatalf("relative = %q", got)
	}
}

// A MULTI-FILE DOCUMENT: the layers panel names the files an include
// wrote into relative to the entry, and the ladder's rungs sort by
// file and then by column.
func TestViewOverIncludedFiles(t *testing.T) {
	dir := t.TempDir()
	viewTempDoc(t, dir, "lib/base.aon", "a: {x: **1 & integer, y: 2}\n")
	entry := viewTempDoc(t, dir, "entry.aon", "@\"./lib/base.aon\"\na: {x: *2 & integer, z: 3}\n")
	src, _ := os.ReadFile(entry)
	a := NewWithBase(dir)
	a.File = entry
	a.Trust = &TrustOptions{IncludeRoot: dir}

	layers := a.View(string(src), &ViewOptions{Kind: "layers"})
	if "rendered" != layers.Verdict {
		t.Fatalf("layers = %+v", layers)
	}
	// The included file is named relative to the entry, in the host's
	// own separator.
	if !strings.Contains(*layers.Text, "# layers  file=entry.aon  documents=2") ||
		!strings.Contains(*layers.Text, filepath.Join("lib", "base.aon")) {
		t.Fatalf("layers text:\n%s", *layers.Text)
	}
	if r := a.View(string(src), &ViewOptions{Kind: "layers", MaxRows: 1}); "error" != r.Verdict ||
		"view_rows_exceeded" != r.Errors[0].Code {
		t.Fatalf("layers max rows = %+v", r)
	}

	ladder := a.View(string(src), &ViewOptions{Kind: "ladder", At: "$.a.x"})
	if "rendered" != ladder.Verdict {
		t.Fatalf("ladder = %+v", ladder)
	}
	want := "c0[\"**1<br/>pref | base.aon:1:8\"]\n  c1[\"*2<br/>pref | entry.aon:2:8\"]\n" +
		"  c2[\"integer<br/>literal | entry.aon:2:13\"]\n  c3[\"integer<br/>literal | base.aon:1:14\"]"
	if !strings.Contains(*ladder.Text, want) {
		t.Fatalf("ladder text:\n%s", *ladder.Text)
	}

	// The poset labels a document by its file, and a further document
	// by its own path.
	other := viewTempDoc(t, dir, "wide.aon", "a: {x: integer, y: integer, z: integer}\n")
	osrc, _ := os.ReadFile(other)
	poset := a.View(string(src), &ViewOptions{Kind: "poset", Profile: "values",
		Docs: []ViewDoc{{Src: string(osrc), Path: other}}})
	if "rendered" != poset.Verdict || !strings.Contains(*poset.Text, "n0[\"entry\"]\n  n1[\"wide\"]\n  n0 --> n1") {
		t.Fatalf("poset = %+v %s", poset, *poset.Text)
	}
}

func TestViewLayersSkipsPathsTheDocumentLacks(t *testing.T) {
	src := "a: {b: 1}\n"
	a := New()
	prov := newProvenance(src, map[string]string{})
	root, _, errs := a.viewLoad(src, prov)
	if nil != errs {
		t.Fatal(errs)
	}
	// A record at a path the document does not have, and a record
	// nothing contributed to: neither is a row.
	ghost := prov.paths["a.b"].conjuncts[0]
	prov.paths["a.ghost"] = &whyPathRecord{conjuncts: []whyContribution{ghost}}
	prov.paths["a.empty"] = &whyPathRecord{}
	loss := []ViewLoss{}
	text, ferrs := drawLayers(prov, root, "", "", 0, 0, "text", "none", 60, &loss)
	if nil != ferrs || strings.Contains(text, "ghost") || strings.Contains(text, "empty") {
		t.Fatalf("layers = %q %v", text, ferrs)
	}
}

func TestViewPosetInjectedVerdicts(t *testing.T) {
	docs := []viewPosetDoc{{src: "a", label: "a"}, {src: "b", label: "b"}, {src: "c", label: "c"}}
	compare := func(g, s viewPosetDoc) (string, string) {
		switch g.label + s.label {
		case "ab", "bc":
			return SubsumeYes, ""
		case "ac":
			return SubsumeNo, "compat_narrowed"
		}
		return SubsumeNo, "compat_narrowed"
	}
	loss := []ViewLoss{}
	text, errs := New().drawPoset(docs, "", "", "mermaid", 60, &loss, compare)
	if nil != errs || !strings.Contains(text, "n1 --> n0\n  n2 --> n1") {
		t.Fatalf("poset = %q %v", text, errs)
	}
	if 1 != len(loss) || "order_intransitive" != loss[0].Code || "c < a" != loss[0].Detail[0] {
		t.Fatalf("loss = %+v", loss)
	}

	bad := []viewPosetDoc{{src: "a", label: "a\nb"}}
	if _, errs = New().drawPoset(bad, "", "", "dot", 60, &loss, compare); nil == errs ||
		"view_line_break" != errs[0].Code {
		t.Fatalf("line break = %v", errs)
	}
}

// ViewDefaultProfile is the kind-to-profile table cmd/aontu reads to
// resolve `--style auto` before the library runs. Its own arms: a kind
// that has profiles, and one that is not a kind at all.
func TestViewDefaultProfile(t *testing.T) {
	for kind, want := range map[string]string{
		"doc": "text", "tree": "text", "matrix": "text",
		"graph": "mermaid", "layer": "text", "sets": "text",
		"layers": "text", "ladder": "mermaid", "poset": "mermaid",
		"nope": "",
	} {
		if got := ViewDefaultProfile(kind); want != got {
			t.Fatalf("ViewDefaultProfile(%q) = %q, want %q", kind, got, want)
		}
	}
}

func TestViewDocWalksAList(t *testing.T) {
	a := New()
	root, _, errs0 := a.viewLoad(`a: [1, {b: "x"}]`, nil)
	if nil != errs0 {
		t.Fatal(errs0)
	}
	loss := []ViewLoss{}
	text, errs := drawDoc(root, "$", 3, "text", "none", 60, &loss)
	if nil != errs {
		t.Fatal(errs)
	}
	want := "$\n└── a\n    ├── 0 1\n    └── 1\n        └── b \"x\""
	if want != text {
		t.Fatalf("list walk =\n%q\nwant\n%q", text, want)
	}
}

File: ts/test/view.test.ts
/* Copyright (c) 2026 Richard Rodger, MIT License */


import { describe, test } from 'node:test'
import * as Assert from 'node:assert'
import * as Fs from 'node:fs'
import * as Os from 'node:os'
import * as Path from 'node:path'

import { view, viewSet } from '../dist/view'
import type { ViewCompare, ViewPosetDoc } from '../dist/view'
import { Provenance } from '../dist/provenance'


const write = (dir: string, name: string, src: string): string => {
  const file = Path.join(dir, name)
  Fs.mkdirSync(Path.dirname(file), { recursive: true })
  Fs.writeFileSync(file, src)
  return file
}


class Ghostly extends Provenance {
  record(path: string[], a: any, b: any, out: any): void {
    super.record(path, a, b, out)
    const rec: any = this.paths.get(path.join('.'))
    if (0 < rec.conjuncts.length && !this.paths.has('a.ghost')) {
      this.paths.set('a.ghost', {
        conjuncts: [rec.conjuncts[0]], made: new Set(), seen: new Set(),
      } as any)
      this.paths.set('a.empty', {
        conjuncts: [], made: new Set(), seen: new Set(),
      } as any)
    }
  }
}


describe('view', () => {

  test('view-over-included-files', () => {
    const dir = Fs.mkdtempSync(Path.join(Os.tmpdir(), 'aontu-view-'))
    write(dir, 'lib/base.aon', 'a: {x: **1 & integer, y: 2}\n')
    const entry = write(dir, 'entry.aon',
      '@"./lib/base.aon"\na: {x: *2 & integer, z: 3}\n')
    const src = Fs.readFileSync(entry, 'utf8')
    const trust = { include: 'root', root: dir } as any

    const layers = view(src, { kind: 'layers', path: entry, trust })
    Assert.equal(layers.verdict, 'rendered')
    Assert.match(layers.text as string,
      /^# layers {2}file=entry\.aon {2}documents=2/)
    // The included file is named relative to the entry, in the host's
    // own separator.
    Assert.ok((layers.text as string).includes(Path.join('lib', 'base.aon')),
      layers.text as string)
    Assert.equal(
      view(src, { kind: 'layers', path: entry, trust, maxRows: 1 })
        .errors?.[0].code,
      'view_rows_exceeded')

    const ladder = view(src, { kind: 'ladder', at: '$.a.x', path: entry, trust })
    Assert.equal(ladder.verdict, 'rendered')
    Assert.ok((ladder.text as string).includes(
      'c0["**1<br/>pref | base.aon:1:8"]\n' +
      '  c1["*2<br/>pref | entry.aon:2:8"]\n' +
      '  c2["integer<br/>literal | entry.aon:2:13"]\n' +
      '  c3["integer<br/>literal | base.aon:1:14"]'), ladder.text as string)

    // The poset labels a document by its file, and a further document
    // by its own path.
    const other = write(dir, 'wide.aon',
      'a: {x: integer, y: integer, z: integer}\n')
    const poset = view(src, {
      kind: 'poset', path: entry, trust, profile: 'values',
      docs: [{ src: Fs.readFileSync(other, 'utf8'), path: other }],
    })
    Assert.equal(poset.verdict, 'rendered', poset.text as string)
    Assert.match(poset.text as string,
      /n0\["entry"\]\n {2}n1\["wide"\]\n {2}n0 --> n1$/)
  })


  // No options at all draws the tree; and a document read from nowhere
  // that includes a file by its absolute path names that file as it is.
  test('view-with-no-options-and-an-absolute-include', () => {
    Assert.deepEqual(view('a: 1'),
      { verdict: 'rendered', kind: 'tree', text: '', loss: [] })
    const dir = Fs.mkdtempSync(Path.join(Os.tmpdir(), 'aontu-view-abs-'))
    const lib = write(dir, 'lib.aon', 'b: 2\n')
    // Spelled with forward slashes: a backslash in a string literal is
    // an escape, and a Windows path is full of them.
    const spelled = lib.split(Path.sep).join('/')
    const r = view(`@"${spelled}"\na: 1\n`, {
      kind: 'layers', trust: { include: 'root', root: dir } as any,
    })
    Assert.equal(r.verdict, 'rendered', JSON.stringify(r.errors))
    Assert.match(r.text as string, /^# layers {2}file=- {2}documents=2/)
    Assert.ok((r.text as string).includes('lib.aon'), r.text as string)
  })


  test('view-layers-skips-paths-the-document-lacks', () => {
    const r = view('a: {b: 1}', { kind: 'layers' }, { provenance: () => new Ghostly() })
    Assert.equal(r.verdict, 'rendered', JSON.stringify(r.errors))
    Assert.doesNotMatch(r.text as string, /ghost|empty/)

    // The same seam through a VIEW DOCUMENT: its one evaluation is the
    // instrumented one, so a layers figure it declares reads this
    // recorder rather than a second run's.
    const set = viewSet('a: {b: 1}\nviews: {l: {kind: layers, out: "l.txt"}}',
      { views: '$.views' }, { provenance: () => new Ghostly() })
    Assert.equal(set.verdict, 'rendered', JSON.stringify(set.errors))
    Assert.doesNotMatch(set.views[0].text as string, /ghost|empty/)
  })


  // A document that does not PARSE has no figure either: the parse
  // failure lands on the collecting context, and is the answer.
  test('view-refuses-a-document-that-does-not-parse', () => {
    const r = view('a: ]', { kind: 'matrix' })
    Assert.equal(r.verdict, 'error')
    Assert.equal(r.errors?.[0].code, 'syntax')
    Assert.deepEqual(r.loss, [])
  })


  test('view-set-needs-the-path-of-its-declarations', () => {
    for (const opts of [undefined, { views: '' }]) {
      const r = viewSet('a: 1', opts)
      Assert.equal(r.verdict, 'error')
      Assert.deepEqual(r.views, [])
      Assert.equal(r.errors?.[0].code, 'view_document_shape')
    }
  })


  test('view-poset-injected-verdicts', () => {
    const compare: ViewCompare = (g: ViewPosetDoc, s: ViewPosetDoc) => {
      const pair = g.label + s.label
      return 'ab' === pair || 'bc' === pair
        ? { verdict: 'subsumes', code: '' }
        : { verdict: 'does_not_subsume', code: 'compat_narrowed' }
    }
    const docs = [{ src: 'b', name: 'b' }, { src: 'c', name: 'c' }]
    const r = view('a', { kind: 'poset', docs, path: 'a.aon' }, { compare })
    Assert.equal(r.verdict, 'lossy')
    Assert.match(r.text as string, /n1 --> n0\n {2}n2 --> n1$/)
    Assert.deepEqual(r.loss,
      [{ code: 'order_intransitive', count: 1, detail: ['c < a'] }])

    const bad = view('a', {
      kind: 'poset', docs: [{ src: 'b', name: 'b\nc' }], path: 'a.aon',
    }, { compare })
    Assert.equal(bad.verdict, 'error')
    Assert.equal(bad.errors?.[0].code, 'view_line_break')
  })
})
